iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/builtin/relational/tvf/PatternMatchTableFunction.java

```diff
@@ -0,0 +1,273 @@
+            .build();
+
+    return TableFunctionAnalysis.builder()
+        .properColumnSchema(properColumnSchema)
+        .requireRecordSnapshot(false)
+        .requiredColumns(
+            TBL_PARAM,
+            Arrays.asList(requiredTimeIndex, requiredDataIndex)) // the 0th column is time
+        .handle(handle)
+        .build();
+  }
+
+  @Override
+  public TableFunctionHandle createTableFunctionHandle() {
+    return new MapTableFunctionHandle();
+  }
+
+  @Override
+  public TableFunctionProcessorProvider getProcessorProvider(
+      TableFunctionHandle tableFunctionHandle) {
+    String pattern =
+        (String) ((MapTableFunctionHandle) tableFunctionHandle).getProperty(PATTERN_PARAM);
+    Double smoothValue =
+        (Double) ((MapTableFunctionHandle) tableFunctionHandle).getProperty(SMOOTH_PARAM);
+    Double threshold =
+        (Double) ((MapTableFunctionHandle) tableFunctionHandle).getProperty(THRESHOLD_PARAM);
+    Double widthLimit =
+        (Double) ((MapTableFunctionHandle) tableFunctionHandle).getProperty(WIDTH_PARAM);
+    Double heightLimit =
+        (Double) ((MapTableFunctionHandle) tableFunctionHandle).getProperty(HEIGHT_PARAM);
+    boolean isPatternFromOrigin =
+        (Boolean) ((MapTableFunctionHandle) tableFunctionHandle).getProperty(PATTERN_SOURCE_PARAM);
+
+    QetchAlgorthm qetchAlgorthm = new QetchAlgorthm();
+    qetchAlgorthm.setThreshold(threshold);
+    qetchAlgorthm.setSmoothValue(smoothValue);
+    qetchAlgorthm.setHeightLimit(heightLimit);
+    qetchAlgorthm.setWidthLimit(widthLimit);
+    qetchAlgorthm.setIsPatternFromOrigin(isPatternFromOrigin);
+    qetchAlgorthm.parsePattern2Automaton(pattern);
+
+    return new TableFunctionProcessorProvider() {
+      @Override
+      public TableFunctionDataProcessor getDataProcessor() {
+        return new ShapeMatchDataProcessor(qetchAlgorthm);
+      }
+    };
+  }
+
+  private static class ShapeMatchDataProcessor implements TableFunctionDataProcessor {
+
+    private QetchAlgorthm qetchAlgorthm;
+
+    public ShapeMatchDataProcessor(QetchAlgorthm qetchAlgorthm) {
+      this.qetchAlgorthm = qetchAlgorthm;
+    }
+
+    @Override
+    public void process(
+        Record input,
+        List<ColumnBuilder> properColumnBuilders,
+        ColumnBuilder passThroughIndexBuilder) {
+
+      double time = input.getLong(0);
+      // need to judge the type of the value and trans it to double
+      double value;
+      if (input.getDataType(1) == Type.INT32) {
+        value = (double) input.getInt(1);
+      } else if (input.getDataType(1) == Type.INT64) {
+        value = (double) input.getLong(1);
+      } else if (input.getDataType(1) == Type.FLOAT) {
+        value = (double) input.getFloat(1);
+      } else if (input.getDataType(1) == Type.DOUBLE) {
+        value = (double) input.getDouble(1);
+      } else {
+        throw new UDFException("Unsupported data type for value column: " + input.getDataType(1));
+      }
```

Comment:
```suggestion
double value = input.getDouble(1);
```
other types can also call getDouble

Reply:
else branch won't be reached any more